File: solutions/lua/bank-account/1/bank-account.lua
local BankAccount = {}

BankAccount.new = function()
        
    local account = {_balance = 0, _opened = true}
        
    account.balance = function(self)
        return self._balance
    end

    account.deposit = function(self, amount)
        assert(self._opened, "can't deposit in a closed account")
        assert(amount > 0, "deposit must be positive")
        self._balance = self._balance + amount
    end

    account.withdraw = function(self, amount)
        assert(self._opened, "can't withdraw from a closed account")
        assert(amount > 0, "withdraw must be positive")
        assert(self._balance - amount > 0, "not enough funds for withdrawal")
        self._balance = self._balance - amount
    end

    account.close = function(self)
        self._opened = false
    end
    
    return account
end

return BankAccount

File: solutions/javascript/bank-account/1/bank-account.js

export class BankAccount {
  
  constructor() {
    this._balance = 0;
    this._active = false;
  }

  open() {
    if (this._active) throw new ValueError('Account is already active');
    this._active = true;
  }

  close() {
    if (!this._active) throw new ValueError('Account is inactive');
    this._active = false;
    this._balance = 0;
  }

  deposit(amount) {
    if (!this._active) throw new ValueError('Account is inactive');
    if (amount < 0) throw new ValueError('Deposit amount must be positive');
    this._balance += amount;
  }

  withdraw(amount) {
    if (!this._active) throw new ValueError('Account is inactive');
    if (amount < 0) throw new ValueError('Withdrawal amount must be positive');
    if (amount > this._balance) throw new ValueError('Unsufficient funds for withdrawal');
    this._balance -= amount;
  }

  get balance() {
    if (!this._active) throw new ValueError('Account is inactive');
    return this._balance;
  }
}

export class ValueError extends Error {
  constructor() {
    super('Bank account error');
  }
}
